### ISSUE TITLE
Recursive optional dependencies lead to incorrect requirements files

### ISSUE THREAD

**Issue by dhdaines** [juftin/hatch-pip-compile#78]
A not entirely documented (https://github.com/pypa/pip/issues/11296) but extremely useful feature of `pip` since version 21.2 is that `optional-dependencies` groups can depend on each other:  https://hynek.me/articles/python-recursive-optional-dependencies/

For example if your package is `mypkg` and it has an optional REST API which you could run with `uvicorn` or some other server, which you run from the `uvicorn` environment in hatch, and for which you'd like to lock dependencies, you could do this:
```toml
[project.optional-dependencies]
api = [
    "fastapi"
]
uvicorn = [
     "mypkg[api]",
     "uvicorn",
]
[tool.hatch.envs.uvicorn]
features = [ "uvicorn" ]
type = "pip-compile"
```
Unfortunately this doesn't really work with `pip-compile` or `hatch-pip-compile`, because you will end up with this in `requirements/requirements-uvicorn.txt`:
```
mypkg==1.0.0
    # via hatch.envs.uvicorn
```
Oh noes!  It went off to PyPI and found that yes, indeed, there is a pakcage there already called `mypkg` and proceeded to add it as a dependency to the requirements file.  That's definitely not what you want!

Note that if you run `pip-compile --extra uvicorn` in the environment you'll get this instead:
```
mypkg[api] @ file:///my/local/path/to/mypkg
    # via file:///my/local/path/to/mypkg
```
Which is not really ideal but definitely better than pulling in some possibly unrelated or out-of-date package from PyPI.

Responsibility for fixing this might be partly in `pip-tools` if there isn't a way for `hatch-pip-compile` to get it to do the right thing...

**Comment by dhdaines** [comment]
At first glance, adding `--unsafe-package my-pkg --no-allow-unsafe` to the `pip-compile` arguments will achieve this, based on https://github.com/jazzband/pip-tools/issues/2002#issuecomment-1786936141

**Comment by juftin** [comment]
This is an interesting one.

`hatch-pip-compile` is an [EnvironmentPlugin](https://hatch.pypa.io/1.1/plugins/environment/) - so at "pip-compile time" it references an environment property, [dependencies](https://hatch.pypa.io/1.1/plugins/environment/#hatch.env.plugin.interface.EnvironmentInterface.dependencies), and sends that over to `pip-compile` as a temporary `requirements.in` file. 

So in a perfect world for your example, the `dependencies` property should return `['fastapi', 'uvicorn']` - but instead it's returning `['mypkg[api]', 'uvicorn']`. 

I can try to add some logic here that detects recursive optional dependencies, gathers all of the ultimate dependencies, and injects them into pip-compile - but IMO an ideal fix would be upstream on the `dependencies` / `dependencies_complex` properties. @ofek I'm curious what you're doing to handle this scenario where optional dependency groups refer to other groups. Is it this a reasonable request for hatch to change the behavior of `dependencies` / `dependencies_complex` for these cases?

**Comment by ofek** [comment]
Yes this is an oversight https://github.com/pypa/hatch/issues/1347

**Comment by dhdaines** [comment]
> At first glance, adding `--unsafe-package mypkg --no-allow-unsafe` to the `pip-compile` arguments will do the right thing, based on [jazzband/pip-tools#2002 (comment)](https://github.com/jazzband/pip-tools/issues/2002#issuecomment-1786936141)

Some further information, this workaround unfortunately doesn't work correctly with `hatch-pip-compile` (by adding `--unsafe-package` to `pip-compile-args`), because (perhaps for the reasons in previous comments) it ends up referring to to the upstream `mypkg[api]`  rather than the local one.  So you still get the wrong transitive dependencies (or lack thereof) even if you don't get the incorrect self-dependency.

I think this is what you're talking about [https://github.com/juftin/hatch-pip-compile/issues/78#issuecomment-2028151851](here)?

**Comment by dhdaines** [comment]
Thanks for looking into this and all the examples and explanation!  For the moment we can just workaround by not using recursive dependencies for the particular environment that uses `pip-compile`.

**Comment by juftin** [comment]
Will be resolved upstream by https://github.com/pypa/hatch/pull/1387

**Comment by ofek** [comment]
Can you please confirm that this works on the master branch of Hatch?

**Comment by juftin** [comment]
> Can you please confirm that this works on the master branch of Hatch?

Yep, confirmed working. Really nice work @ofek 🙇 

<details><summary>pyproject.toml</summary>
<p>

```toml
[project]
name = "mypkg1234"
version = "1.0.0"

[project.optional-dependencies]
api = [
    "fastapi"
]
uvicorn = [
     "mypkg1234[api]",
     "uvicorn",
]

[tool.hatch.envs.uvicorn]
features = [ "uvicorn" ]
type = "pip-compile"
```


</p>
</details> 

<details><summary>requirements/requirements-uvicorn.txt</summary>
<p>


```text
#
# This file is autogenerated by hatch-pip-compile with Python 3.12
#
# - fastapi
# - uvicorn
#

annotated-types==0.6.0
    # via pydantic
anyio==4.3.0
    # via starlette
click==8.1.7
    # via uvicorn
fastapi==0.110.2
    # via hatch.envs.uvicorn
h11==0.14.0
    # via uvicorn
idna==3.7
    # via anyio
pydantic==2.7.0
    # via fastapi
pydantic-core==2.18.1
    # via pydantic
sniffio==1.3.1
    # via anyio
starlette==0.37.2
    # via fastapi
typing-extensions==4.11.0
    # via
    #   fastapi
    #   pydantic
    #   pydantic-core
uvicorn==0.29.0
    # via hatch.envs.uvicorn
```

</p>
</details>

**Comment by ofek** [comment]
Awesome, thank you! Just FYI while I have you here, the next release will begin shipping UV directly as a dependency.

**Comment by juftin** [comment]
> Awesome, thank you! Just FYI while I have you here, the next release will begin shipping UV directly as a dependency.

I saw `uv` fly by while I was installing from master, that is awesome!

**Comment by ovidiu-eremia** [comment]
Hi @juftin,

First, thank you for maintaining this useful plugin for us, hatch users!

Second, I also have this issue now with hatch-pip-compile and latest hatch version.
Could you please let us know if this will be fixed also downstream in this plugin?

Thanks!
